### PR TITLE
feat(eigen-client-m0-implementation): concurrent da dispatcher

### PR DESCRIPTION
**Pull request by juan518munoz**
## What ❔

Implement concurrent data availability dispatcher, to greatly reduce dispersing time when there are many blobs due dispersing.

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zkstack dev fmt` and `zkstack dev lint`.

### CODE REVIEW

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 1]
```diff
@@ -27,158 +29,312 @@ impl DataAvailabilityDispatcher {
...
+            loop {
+                if *stop_receiver_clone.borrow() {
+                    break;
+                }
+
+                let mut conn = pool_clone.connection_tagged("da_dispatcher").await?;
+
+                // TODO: this query might always return the same blob if the blob is not included
+                // we should probably change the query to return all blobs that are not included
+                let blob_info = conn
```
[the remark this answers is not in the document]
Comment (juan518munoz) [reply core/node/da_dispatcher/src/da_dispatcher.rs]: Corretct, added it!

File: core/lib/config/src/configs/da_dispatcher.rs [hunk 2]
```diff
@@ -6,6 +6,7 @@ pub const DEFAULT_POLLING_INTERVAL_MS: u32 = 5000;
 pub const DEFAULT_MAX_ROWS_TO_DISPATCH: u32 = 100;
 pub const DEFAULT_MAX_RETRIES: u16 = 5;
 pub const DEFAULT_USE_DUMMY_INCLUSION_DATA: bool = false;
+pub const DEFAULT_MAX_CONCURRENT_REQUESTS: u32 = 100;
```
Comment (juanbono) [inline core/lib/config/src/configs/da_dispatcher.rs:9]: I think this should be a configurable parameter 🤔

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 3]
```diff
@@ -27,158 +29,307 @@ impl DataAvailabilityDispatcher {
...
+            _ = subtasks => {},
+        }
+        Ok(())
+    }
 
-            let subtasks = futures::future::join(
-                async {
-                    if let Err(err) = self.dispatch().await {
-                        tracing::error!("dispatch error {err:?}");
+    async fn dispatch_batches(&self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:65]: why the unit Error?

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 4]
```diff
@@ -27,158 +29,307 @@ impl DataAvailabilityDispatcher {
...
+                        batch.l1_batch_number,
+                        batch.pubdata.len(),
+                    );
+
+                    Ok::<(), anyhow::Error>(())
+                });
+                spawned_requests.push(request);
+            }
+            join_all(spawned_requests).await;
+            Ok::<(), anyhow::Error>(())
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:209]: same here

File: core/node/da_dispatcher/src/da_dispatcher.rs [hunk 5]
```diff
@@ -27,158 +29,307 @@ impl DataAvailabilityDispatcher {
...
-        }
-        METRICS
-            .last_included_l1_batch
-            .set(blob_info.l1_batch_number.0 as usize);
-
-        tracing::info!(
-            "Received an inclusion data for a batch_number: {}, inclusion_latency_seconds: {}",
-            blob_info.l1_batch_number,
-            inclusion_latency.num_seconds()
+    async fn inclusion_poller(&self, stop_receiver: Receiver<bool>) -> anyhow::Result<()> {
```
Comment (juanbono) [inline core/node/da_dispatcher/src/da_dispatcher.rs:219]: same here

**Review by juanbono — APPROVED** [review]
I think it's more useful to not return any

**Review by gianbelinche — APPROVED** [review]
LGTM

**Pull request merged**